public/app/features/dashboard/components/PanelEditor/state/actions.ts

```diff
@@ -106,17 +106,19 @@ export function skipPanelUpdate(modifiedPanel: PanelModel, panelToUpdate: PanelM
   return false;
 }
 
-export function exitPanelEditor(): ThunkResult<void> {
+export function exitPanelEditor(forceRefresh = false): ThunkResult<void> {
   return async (dispatch, getStore) => {
     const dashboard = getStore().dashboard.getModel();
     const { getPanel, getSourcePanel, shouldDiscardChanges } = getStore().panelEditor;
     const panel = getPanel();
+    const sourcePanel = getSourcePanel();
+    const libraryPanelChanged = panel.libraryPanel?.uid !== sourcePanel.libraryPanel?.uid;
 
     if (dashboard) {
       dashboard.exitPanelEditor();
     }
 
-    if (panel.hasChanged && !shouldDiscardChanges) {
+    if ((forceRefresh || panel.hasChanged || libraryPanelChanged) && !shouldDiscardChanges) {
```

Comment:
Something that feels off here, has not panel library changed in this scenario? 


forceRefresh is also a bit of confusing name , as refresh is also what we call when we any to execute queries (refresh called on dashboard and panels)

Reply:
`forceRefresh` is true when the we leave the panel editor, the panel being edited is a library panel, a property of that library panel has been changed, and "confirm" is pressed on the displayed modal. `panel.hasChanged` doesn't get set when we actually _replace_ the library panel from the panel editor though, because we don't want the modal to pop up in that case. We still want the panel to be refreshed though, so that's the reason for the `libraryPanelChanged` condition. In hindsight a better name might be `libraryPanelReplaced`.
Agree about `forceRefresh`, I'll try and think of a better name.